File: src/translation/tools/translate.rs
use crate::llm::ToolDef;

#[derive(serde::Deserialize)]
pub struct TranslationItem {
    pub id: String,
    pub translated_text: String,
}

#[derive(serde::Deserialize)]
pub struct Args {
    pub translations: Vec<TranslationItem>,
}

pub fn def() -> ToolDef {
    ToolDef::new(
        "translate",
        "Submit translations for one or more bubbles in a single call.\n\n\
            Behavior:\n\
            - Each item: id (must match a required bubble ID), translated_text (final localized text).\n\
            - Submit as many bubbles as possible per call to minimize round-trips.\n\
            - To revise: include the same id again (latest wins).\n\
            - Do not skip any listed bubble, even if short or SFX.",
        serde_json::json!({
            "type": "object",
            "required": ["translations"],
            "additionalProperties": false,
            "properties": {
                "translations": {
                    "type": "array",
                    "description": "Array of bubble translations",
                    "items": {
                        "type": "object",
                        "required": ["id", "translated_text"],
                        "additionalProperties": false,
                        "properties": {
                            "id": {
                                "type": "string",
                                "description": "Bubble ID exactly as listed (e.g. b0, p0_b1)"
                            },
                            "translated_text": {
                                "type": "string",
                                "description": "Final translated text — no notes or explanations"
                            }
                        }
                    }
                }
            }
        }),
    )
    .strict()
}

File: typoon/translation/tools/translate.py
from __future__ import annotations

from pydantic import BaseModel, Field

from typoon.llm.tool_dec import tool


class TranslationItem(BaseModel):
    id: str = Field(description="Bubble ID exactly as listed (e.g. p0_b0, p0_b1)")
    translated_text: str = Field(description="Final translated text. Empty string for noise.")


class TranslateArgs(BaseModel):
    translations: list[TranslationItem] = Field(description="Array of bubble translations")


@tool(strict=True)
async def translate(args: TranslateArgs) -> str:
    """Submit translations for one or more bubbles in a single call.

    - Submit as many bubbles as possible per call to minimize round-trips.
    - To revise: include the same id again (latest wins).
    - For non-dialogue noise (watermarks, URLs, page numbers, credits, decorative SFX):
      use empty translated_text.
    - Meaning-bearing SFX: translate/adapt naturally.
    """
    return f"ok ({len(args.translations)} bubbles)"
